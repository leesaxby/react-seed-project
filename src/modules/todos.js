import { push } from 'react-router-redux';

const ADD_TODO = 'app/todos/ADD_TODO';
const FETCH_TODOS_SUCCESS = 'app/todos/FETCH_TODOS_SUCCESS';
const UPDATE_DONE_STATUS = 'app/todos/UPDATE_DONE_STATUS';
const CHANGE_FILTER = 'app/todos/CHANGE_FILTER';

const addTodo = (todo) => ({
    type: ADD_TODO,
    payload: todo
});

const updateDoneStatus = (_id, doneStatus) => ({
    type: UPDATE_DONE_STATUS,
    payload: { _id, doneStatus }
});

const fetchTodosSuccess = () => ({
    type: FETCH_TODOS_SUCCESS,
    payload: [
        { _id: 1, text: 'Item One', done: false},
        { _id: 2, text: 'Item Two', done: false},
        { _id: 3, text: 'Item Three', done: false}
    ]
});

const fetchTodos = () => dispatch => Promise.resolve().then(() => dispatch(fetchTodosSuccess()));

const changeFilter = (filter) => dispatch => {
    dispatch(push(filter));
    dispatch({
        type: CHANGE_FILTER,
        payload: filter
    });
};

export {
    ADD_TODO,
    FETCH_TODOS_SUCCESS,
    UPDATE_DONE_STATUS,
    addTodo,
    updateDoneStatus,
    fetchTodos,
    changeFilter
};

const initialState = {
    listItems: [],
    filter: 'ACTIVE'
};

const todo = (state, action) => {
    switch (action.type) {
        case ADD_TODO: {
            const ids = state.listItems.map(todo => todo._id);
            const newId = ids.length ? Math.max(...ids) + 1 : 0;
            return Object.assign({}, action.payload, { _id: newId });
        }
        case UPDATE_DONE_STATUS: {
            const listItems = Object.assign(state.listItems);
            const itemIndex = listItems.findIndex(({ _id }) => _id === action.payload._id);
            const newListItem = Object.assign({}, listItems[itemIndex], {
                done: action.payload.doneStatus
            });

            return { [itemIndex]: newListItem };
        }
        default:
            return state;
    }
};

export default function todos(state = initialState, action) {
    switch (action.type) {
        case ADD_TODO: {
            return Object.assign({}, state, {
                listItems: [
                    ...state.listItems,
                    todo(state, action)
                ]
            });
        }
        case FETCH_TODOS_SUCCESS:
            return Object.assign({}, state, {
                listItems: action.payload
            });
        case UPDATE_DONE_STATUS: {
            return Object.assign({}, state, {
                listItems: Object.assign([], state.listItems, todo(state, action))
            });
        }
        case CHANGE_FILTER:
            return Object.assign({}, state, {
                filter: action.payload
            });
        default:
            return state;
    }
}