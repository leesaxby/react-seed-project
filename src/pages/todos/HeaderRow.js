import React from 'react';
import { Row, Col, Image } from 'react-bootstrap';

import listImage from 'Images/list.png';

export default function HeaderRow() {
	return (
	    <Row>
	        <Col sm={1} lg={1} smOffset={4} lgOffset={5}>
	            <Image src={ listImage }
	                   alt="List Image"
	                   style={{ 'height': '200px',
	                            'marginBottom': '20px' }}/>
								asdasdsdasd
	        </Col>
	    </Row>
	);
}
