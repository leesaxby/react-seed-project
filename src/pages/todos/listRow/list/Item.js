import React from 'react';
import PropTypes from 'prop-types';
import { ListGroupItem, FormControl, Glyphicon } from 'react-bootstrap';

export default function Item({ item, onToggleDone }) {
    const toggleDone = () => {
        onToggleDone(item);
    };

    const labelId = `todoItem${item._id}Label`;
    const aria = {
        role: 'button',
        tabIndex: 0,
        onKeyDown: e => {
            if(e.keyCode === 13) {
                toggleDone();
            }
        },
        'aria-labelledby': labelId
    };
            
    return (
        <ListGroupItem { ...aria } onClick={ toggleDone }
                       data-test-id="todo-item">

            <FormControl.Static id={ labelId } style={{ 'textDecoration': item.done ? 'line-through' : '' }}>

                <Glyphicon glyph="glyphicon glyphicon-ok"
                           style={{ 'marginRight': '15px' }}/>
                { item.text }

            </FormControl.Static>

        </ListGroupItem>
    );
}

Item.propTypes = {
    item: PropTypes.shape({
        _id: PropTypes.number.isRequired,
        text: PropTypes.string,
        done: PropTypes.bool.isRequired
    }),
    onToggleDone: PropTypes.func.isRequired
};
