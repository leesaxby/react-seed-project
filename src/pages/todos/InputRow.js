import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { push } from 'react-router-redux';
import { Row, Col } from 'react-bootstrap';
import { addTodo } from 'Modules/todos';
import Form from './inputRow/Form';
import Filter from './inputRow/Filter';

export const InputRow = ({ updateFilter, addTodo, filter }) => (
    <Row>
        <Col sm={8} lg={8} smOffset={1} lgOffset={2}>
            <Form onAddItem={ addTodo } />
        </Col>
        <Col sm={3} lg={2}>
            <Filter filter={ filter } onUpdateFilter={ updateFilter } />
        </Col>
    </Row>
);

InputRow.propTypes = {
    filter: PropTypes.oneOf([ 'ACTIVE', 'DONE' ]),
    addTodo: PropTypes.func.isRequired,
    updateFilter: PropTypes.func.isRequired,
};

const mapStateToProps = ({ todos: { filter } }) => ({ filter });

const mapDispatchToProps = (dispatch) => {
    return {
        updateFilter: (filter) => dispatch(push(filter)),
        addTodo: (todo) => dispatch(addTodo({ text: todo, done: false })),
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(InputRow);