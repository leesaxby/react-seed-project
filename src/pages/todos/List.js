import React from 'react';
import PropTypes from 'prop-types';
import { List } from '@material-ui/core';
import Item from './list/Item';

export const TodoList = ({ listItems, onItemSelect }) => (
    <List>
        {
            listItems.map((item) => (
                <Item
                    key={item.id}
                    item={item}
                    onItemSelect={onItemSelect} />
            ))
        }
    </List>
);

TodoList.defaultProps = {
    listItems: [],
};

TodoList.propTypes = {
    listItems: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.number.isRequired,
            text: PropTypes.string,
            done: PropTypes.bool.isRequired,
        }),
    ),
    onItemSelect: PropTypes.func.isRequired,
};

export default TodoList;
