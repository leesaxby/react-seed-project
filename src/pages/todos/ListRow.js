import React from 'react';
import List from './listRow/List';

export default function ListRow() {
    return (
        <List />
    );
}
