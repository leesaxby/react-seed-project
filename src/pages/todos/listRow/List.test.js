import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';

import ListContainer, { List } from './List';
import Item from './list/Item';
import { updateDoneStatus } from 'Modules/todos';

describe('A List', () => {

	function setup() {
		const props = {
		    listItems: [{
		    	_id: 1,
		    	done: true,
		    }, {
		    	_id: 2,
		    	done: false,
		    }],
		    onItemClick: jest.fn(),
		};
		const wrapper = shallow(<List { ...props }/>);
	    return {
	        props,
	        wrapper,
	    };
	}

	it('renders component', () => {
		const { props } = setup();
		const tree = renderer.create(
			<List { ...props } />
		).toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('triggers a callback when an Item is clicked', () => {
		const { props, wrapper } = setup();

        expect(props.onItemClick.mock.calls).toHaveLength(0);

        wrapper.find(Item).at(0).simulate('click');
        expect(props.onItemClick.mock.calls).toHaveLength(1);

        wrapper.find(Item).at(1).simulate('click');
        expect(props.onItemClick.mock.calls).toHaveLength(2);
	});
});

describe('A List Container', () => {

	function setup() {
		const initialState = {
			todos: {
				filter: 'ACTIVE',
				listItems: [{
					_id: 1,
					done: true,
				}, {
					_id: 2,
					done: false,
				}],
			},
		};
		const store = configureStore()(initialState);

		const wrapper = shallow(<ListContainer store={store} />);

		return {
			initialState,
			store,
			wrapper,
		};
	}

	it('renders component', () => {
		const { store } = setup();
		const tree = renderer.create(
			<ListContainer store={store} />
		).toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('correctly filters todos depending on the filter state', () => {
		const { initialState } = setup();
		const newState = {
			todos: {
				filter: 'DONE',
				listItems: initialState.todos.listItems,
			},
		};
		const store = configureStore()(newState);

		const tree = renderer.create(
			<ListContainer store={store} />
		).toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('maps the right portion of the state to List props', () => {
		const { wrapper, initialState } = setup();

		expect(wrapper.find(List).prop('listItems')).toEqual(initialState.todos.listItems.filter((item) => !item.done));
	});

	it('dispatches the right actions from List props', () => {
		const { wrapper, initialState, store } = setup();

		expect(store.getActions()).toHaveLength(0);

		const clickedItem = initialState.todos.listItems[0];

		wrapper.prop('onItemClick')(clickedItem);

		expect(store.getActions()).toHaveLength(1);
		expect(store.getActions()[0]).toEqual(updateDoneStatus(clickedItem._id, !clickedItem.done));
	});

});