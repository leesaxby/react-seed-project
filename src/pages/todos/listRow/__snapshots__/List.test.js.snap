// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A List Container correctly filters todos depending on the filter state 1`] = `
<ul
  className="MuiList-root MuiList-padding"
>
  <li
    aria-disabled={false}
    aria-labelledby="todoItemButton1Label"
    className="MuiButtonBase-root MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-button"
    data-test-id="todo-item"
    id="todoItemButton1Label"
    label="todoItemButton1Label"
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="button"
    style={
      Object {
        "textDecoration": "line-through",
      }
    }
    tabIndex={0}
  >
    <label
      htmlFor="checkboxLabelId"
    >
      <span
        aria-disabled={false}
        aria-label="todoItemCheckbox1Label"
        className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-25 MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-checked-26 Mui-checked MuiIconButton-colorSecondary"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
          <input
            checked={true}
            className="PrivateSwitchBase-input-28"
            data-indeterminate={false}
            id="todoItemCheckbox1Label"
            onChange={[Function]}
            tabIndex={-1}
            type="checkbox"
          />
        </span>
      </span>
    </label>
  </li>
</ul>
`;

exports[`A List Container renders component 1`] = `
<ul
  className="MuiList-root MuiList-padding"
>
  <li
    aria-disabled={false}
    aria-labelledby="todoItemButton2Label"
    className="MuiButtonBase-root MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-button"
    data-test-id="todo-item"
    id="todoItemButton2Label"
    label="todoItemButton2Label"
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="button"
    style={
      Object {
        "textDecoration": "",
      }
    }
    tabIndex={0}
  >
    <label
      htmlFor="checkboxLabelId"
    >
      <span
        aria-disabled={false}
        aria-label="todoItemCheckbox2Label"
        className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-25 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
          <input
            checked={false}
            className="PrivateSwitchBase-input-28"
            data-indeterminate={false}
            id="todoItemCheckbox2Label"
            onChange={[Function]}
            tabIndex={-1}
            type="checkbox"
          />
        </span>
      </span>
    </label>
    <span
      className="MuiTouchRipple-root"
    />
  </li>
</ul>
`;

exports[`A List renders component 1`] = `
<ul
  className="MuiList-root MuiList-padding"
>
  <li
    aria-disabled={false}
    aria-labelledby="todoItemButton1Label"
    className="MuiButtonBase-root MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-button"
    data-test-id="todo-item"
    id="todoItemButton1Label"
    label="todoItemButton1Label"
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="button"
    style={
      Object {
        "textDecoration": "line-through",
      }
    }
    tabIndex={0}
  >
    <label
      htmlFor="checkboxLabelId"
    >
      <span
        aria-disabled={false}
        aria-label="todoItemCheckbox1Label"
        className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-25 MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-checked-26 Mui-checked MuiIconButton-colorSecondary"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
          <input
            checked={true}
            className="PrivateSwitchBase-input-28"
            data-indeterminate={false}
            id="todoItemCheckbox1Label"
            onChange={[Function]}
            tabIndex={-1}
            type="checkbox"
          />
        </span>
      </span>
    </label>
  </li>
  <li
    aria-disabled={false}
    aria-labelledby="todoItemButton2Label"
    className="MuiButtonBase-root MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-button"
    data-test-id="todo-item"
    id="todoItemButton2Label"
    label="todoItemButton2Label"
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="button"
    style={
      Object {
        "textDecoration": "",
      }
    }
    tabIndex={0}
  >
    <label
      htmlFor="checkboxLabelId"
    >
      <span
        aria-disabled={false}
        aria-label="todoItemCheckbox2Label"
        className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-25 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
          <input
            checked={false}
            className="PrivateSwitchBase-input-28"
            data-indeterminate={false}
            id="todoItemCheckbox2Label"
            onChange={[Function]}
            tabIndex={-1}
            type="checkbox"
          />
        </span>
      </span>
    </label>
  </li>
</ul>
`;
